src/query/plan/rule_based_planner.hpp

```diff
@@ -567,6 +565,66 @@ class RuleBasedPlanner {
     return std::make_unique<plan::Foreach>(std::move(input_op), std::move(op), foreach->named_expression_->expression_,
                                            symbol);
   }
+
+  std::unique_ptr<LogicalOperator> GenFilters(std::unique_ptr<LogicalOperator> last_op,
+                                              const std::unordered_set<Symbol> &bound_symbols, Filters &filters,
+                                              AstStorage &storage, const SymbolTable &symbol_table) {
+    std::unique_ptr<LogicalOperator> pattern_filter =
+        ExtractPatternFilters(filters, storage, symbol_table, bound_symbols);
+    auto *filter_expr = impl::ExtractFilters(bound_symbols, filters, storage);
+    if (filter_expr) {
+      last_op = std::make_unique<Filter>(std::move(last_op), std::move(pattern_filter), filter_expr);
+    }
+    return last_op;
+  }
+
+  std::unique_ptr<LogicalOperator> ExtractPatternFilters(const Filters &filters, AstStorage &storage,
+                                                         const SymbolTable &symbol_table,
+                                                         const std::unordered_set<Symbol> &bound_symbols) {
+    for (const auto &filter : filters) {
+      if (filter.type != FilterInfo::Type::Pattern) {
+        continue;
+      }
+
+      if (const auto *exists = utils::Downcast<Exists>(filter.expression)) {
+        return MakeExistsFilter(*exists, storage, symbol_table, bound_symbols);
```

Comment:
Dumb question, but what if in filters we have more than one Exists? Shouldn't we process that too?

Reply:
This just takes the first `Exists`, right?

Reply:
Addressed in following PRs